src/stream/src/executor/hash_join.rs

```diff
@@ -902,6 +898,16 @@ impl<K: HashKey, S: StateStore, const T: JoinTypePrimitive> HashJoinExecutor<K,
                 for output_idx in output_indices {
                     watermarks_to_emit.push(selected_watermark.clone().with_idx(*output_idx));
                 }
+                // State cleaning
+                if idx == 0 {
+                    side_update
+                        .ht
+                        .update_watermark(selected_watermark.val.clone());
+                    side_match
+                        .ht
+                        .update_watermark(selected_watermark.val.clone());
+                }
+                self.current_watermarks[idx] = Some(selected_watermark);
```

Comment:
The watermarks on join keys are recorded here but seems never used. But still it is fine to me.

Reply:
> The watermarks on join keys are recorded here but seems never used. But still it is fine to me.

After a second thought I found that they are totally unnecessary, so I have just removed them.